File: server/debug_logger.py
"""
Debug Logging Utility for Python Backend

Centralized debug logging based on environment variables.
Each debug category can be enabled/disabled via .env file.

Usage:
    from debug_logger import debug_log, is_debug_enabled
    debug_log('OCR_SERVER', f'Processing image: {filename}')
    if is_debug_enabled('MATH_SOLVER'):
        # Expensive debug operation
        pass
"""

import os
from typing import Any, Optional
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

# Check if master debug is enabled
DEBUG_ALL = os.getenv('DEBUG_ALL', 'false').lower() == 'true'

# Debug flag mappings
DEBUG_FLAGS = {
    # OCR Server
    'OCR_SERVER': os.getenv('DEBUG_OCR_SERVER', 'false').lower() == 'true' or DEBUG_ALL,
    'OCR_SAVE_IMAGES': os.getenv('DEBUG_OCR_SAVE_IMAGES', 'false').lower() == 'true' or DEBUG_ALL,
    'TROCR_MODEL': os.getenv('DEBUG_TROCR_MODEL', 'false').lower() == 'true' or DEBUG_ALL,
    
    # Math Server
    'MATH_SERVER': os.getenv('DEBUG_MATH_SERVER', 'false').lower() == 'true' or DEBUG_ALL,
    'MATH_FAST_PATH': os.getenv('DEBUG_MATH_FAST_PATH', 'false').lower() == 'true' or DEBUG_ALL,
    'MATH_SLOW_PATH': os.getenv('DEBUG_MATH_SLOW_PATH', 'false').lower() == 'true' or DEBUG_ALL,
    'MATH_CLASSIFICATION': os.getenv('DEBUG_MATH_CLASSIFICATION', 'false').lower() == 'true' or DEBUG_ALL,
    'MATH_HISTORY': os.getenv('DEBUG_MATH_HISTORY', 'false').lower() == 'true' or DEBUG_ALL,
    
    # Equation Solving
    'EQUATION_CLASSIFIER': os.getenv('DEBUG_EQUATION_CLASSIFIER', 'false').lower() == 'true' or DEBUG_ALL,
    'SMART_MATH_ENGINE': os.getenv('DEBUG_SMART_MATH_ENGINE', 'false').lower() == 'true' or DEBUG_ALL,
    'FAST_MATH_SOLVER': os.getenv('DEBUG_FAST_MATH_SOLVER', 'false').lower() == 'true' or DEBUG_ALL,
    'PHYSICS_SOLVER': os.getenv('DEBUG_PHYSICS_SOLVER', 'false').lower() == 'true' or DEBUG_ALL,
    
    # Server Health & Monitoring
    'SERVER_HEALTH': os.getenv('DEBUG_SERVER_HEALTH', 'false').lower() == 'true' or DEBUG_ALL,
    'SERVER_STATS': os.getenv('DEBUG_SERVER_STATS', 'false').lower() == 'true' or DEBUG_ALL,
}


def is_debug_enabled(category: str) -> bool:
    """
    Check if a debug category is enabled.
    
    Args:
        category: Debug category name
        
    Returns:
        True if debug is enabled for this category
    """
    return DEBUG_FLAGS.get(category, False)


def debug_log(category: str, *args: Any, **kwargs: Any) -> None:
    """
    Log a debug message if the category is enabled.
    
    Args:
        category: Debug category name
        *args: Arguments to print
        **kwargs: Keyword arguments for print()
    """
    if is_debug_enabled(category):
        print(f"[{category}]", *args, **kwargs)


def debug_warn(category: str, *args: Any, **kwargs: Any) -> None:
    """
    Log a debug warning if the category is enabled.
    
    Args:
        category: Debug category name
        *args: Arguments to print
        **kwargs: Keyword arguments for print()
    """
    if is_debug_enabled(category):
        print(f"[{category}] WARNING:", *args, **kwargs)


def debug_error(category: str, *args: Any, **kwargs: Any) -> None:
    """
    Log a debug error if the category is enabled.
    
    Args:
        category: Debug category name
        *args: Arguments to print
        **kwargs: Keyword arguments for print()
    """
    if is_debug_enabled(category):
        print(f"[{category}] ERROR:", *args, **kwargs)


def get_enabled_debug_flags() -> list[str]:
    """
    Get all enabled debug flags.
    
    Returns:
        List of enabled debug category names
    """
    return [key for key, value in DEBUG_FLAGS.items() if value]


def print_debug_config() -> None:
    """Print debug configuration summary."""
    enabled = get_enabled_debug_flags()
    if not enabled:
        print('[DEBUG] No debug flags enabled')
    elif DEBUG_ALL:
        print('[DEBUG] ALL debug flags enabled')
    else:
        print('[DEBUG] Enabled flags:', ', '.join(enabled))


# Export for easier access
__all__ = [
    'is_debug_enabled',
    'debug_log',
    'debug_warn',
    'debug_error',
    'get_enabled_debug_flags',
    'print_debug_config',
    'DEBUG_FLAGS',
]

File: src/lib/debugLogger.js
/**
 * Debug Logging Utility
 * 
 * Centralized debug logging based on environment variables.
 * Each debug category can be enabled/disabled via .env file.
 * 
 * Usage:
 *   import { debugLog } from './debugLogger';
 *   debugLog('CANVAS', 'Drawing started', { x: 100, y: 200 });
 *   debugLog('MATH_SOLVER', 'Solving equation:', equation);
 */

// Check if master debug is enabled
const DEBUG_ALL = import.meta.env.VITE_DEBUG_ALL === 'true';

// Debug flag mappings
const DEBUG_FLAGS = {
  // Canvas Operations
  CANVAS: import.meta.env.VITE_DEBUG_CANVAS === 'true' || DEBUG_ALL,
  CANVAS_HISTORY: import.meta.env.VITE_DEBUG_CANVAS_HISTORY === 'true' || DEBUG_ALL,
  CANVAS_OPTIMIZER: import.meta.env.VITE_DEBUG_CANVAS_OPTIMIZER === 'true' || DEBUG_ALL,
  
  // Drawing Tools
  BRUSH: import.meta.env.VITE_DEBUG_BRUSH === 'true' || DEBUG_ALL,
  ERASER: import.meta.env.VITE_DEBUG_ERASER === 'true' || DEBUG_ALL,
  SHAPES: import.meta.env.VITE_DEBUG_SHAPES === 'true' || DEBUG_ALL,
  
  // Handwriting Detection & Recognition
  HANDWRITING: import.meta.env.VITE_DEBUG_HANDWRITING === 'true' || DEBUG_ALL,
  EQUALSIGN: import.meta.env.VITE_DEBUG_EQUALSIGN === 'true' || DEBUG_ALL,
  STROKE_ANALYZER: import.meta.env.VITE_DEBUG_STROKE_ANALYZER === 'true' || DEBUG_ALL,
  SPATIAL_CONTEXT: import.meta.env.VITE_DEBUG_SPATIAL_CONTEXT === 'true' || DEBUG_ALL,
  SPATIAL_HASH_GRID: import.meta.env.VITE_DEBUG_SPATIAL_HASH_GRID === 'true' || DEBUG_ALL,
  
  // OCR (Optical Character Recognition)
  OCR: import.meta.env.VITE_DEBUG_OCR === 'true' || DEBUG_ALL,
  OCR_CONFIG: import.meta.env.VITE_DEBUG_OCR_CONFIG === 'true' || DEBUG_ALL,
  OCR_PERFORMANCE: import.meta.env.VITE_DEBUG_OCR_PERFORMANCE === 'true' || DEBUG_ALL,
  LOCAL_OCR: import.meta.env.VITE_DEBUG_LOCAL_OCR === 'true' || DEBUG_ALL,
  OPENROUTER_OCR: import.meta.env.VITE_DEBUG_OPENROUTER_OCR === 'true' || DEBUG_ALL,
  OCR_LEARNING: import.meta.env.VITE_DEBUG_OCR_LEARNING === 'true' || DEBUG_ALL,
  OCR_CACHE: import.meta.env.VITE_DEBUG_OCR_CACHE === 'true' || DEBUG_ALL,
  
  // Math Recognition & Solving
  MATH_DETECTION: import.meta.env.VITE_DEBUG_MATH_DETECTION === 'true' || DEBUG_ALL,
  MATH_SOLVER: import.meta.env.VITE_DEBUG_MATH_SOLVER === 'true' || DEBUG_ALL,
  MATH_ENGINE: import.meta.env.VITE_DEBUG_MATH_ENGINE === 'true' || DEBUG_ALL,
  SMART_MATH_SOLVER: import.meta.env.VITE_DEBUG_SMART_MATH_SOLVER === 'true' || DEBUG_ALL,
  MATH_SIZE_MATCHING: import.meta.env.VITE_DEBUG_MATH_SIZE_MATCHING === 'true' || DEBUG_ALL,
  
  // UI Components
  RESIZE_BOX: import.meta.env.VITE_DEBUG_RESIZE_BOX === 'true' || DEBUG_ALL,
  RECOGNIZE_EQUATION: import.meta.env.VITE_DEBUG_RECOGNIZE_EQUATION === 'true' || DEBUG_ALL,
  MATH_SOLUTION_DISPLAY: import.meta.env.VITE_DEBUG_MATH_SOLUTION_DISPLAY === 'true' || DEBUG_ALL,
  CORRECTION_MODAL: import.meta.env.VITE_DEBUG_CORRECTION_MODAL === 'true' || DEBUG_ALL,
  
  // Symbol & Formula Processing
  SYMBOL_POST_PROCESSING: import.meta.env.VITE_DEBUG_SYMBOL_POST_PROCESSING === 'true' || DEBUG_ALL,
  FORMULA_PATTERNS: import.meta.env.VITE_DEBUG_FORMULA_PATTERNS === 'true' || DEBUG_ALL,
  CONTEXTUAL_CORRECTION: import.meta.env.VITE_DEBUG_CONTEXTUAL_CORRECTION === 'true' || DEBUG_ALL,
  
  // IndexedDB & Storage
  INDEXEDDB: import.meta.env.VITE_DEBUG_INDEXEDDB === 'true' || DEBUG_ALL,
  DB: import.meta.env.VITE_DEBUG_DB === 'true' || DEBUG_ALL,
  
  // Performance & Optimization
  PERFORMANCE: import.meta.env.VITE_DEBUG_PERFORMANCE === 'true' || DEBUG_ALL,
  VIRTUAL_CANVAS: import.meta.env.VITE_DEBUG_VIRTUAL_CANVAS === 'true' || DEBUG_ALL,
};

/**
 * Check if a debug category is enabled
 * @param {string} category - Debug category name
 * @returns {boolean} - True if debug is enabled for this category
 */
export function isDebugEnabled(category) {
  return DEBUG_FLAGS[category] === true;
}

/**
 * Log a debug message if the category is enabled
 * @param {string} category - Debug category name
 * @param {...any} args - Arguments to log
 */
export function debugLog(category, ...args) {
  if (isDebugEnabled(category)) {
    console.log(`[${category}]`, ...args);
  }
}

/**
 * Log a debug warning if the category is enabled
 * @param {string} category - Debug category name
 * @param {...any} args - Arguments to log
 */
export function debugWarn(category, ...args) {
  if (isDebugEnabled(category)) {
    console.warn(`[${category}]`, ...args);
  }
}

/**
 * Log a debug error if the category is enabled
 * @param {string} category - Debug category name
 * @param {...any} args - Arguments to log
 */
export function debugError(category, ...args) {
  if (isDebugEnabled(category)) {
    console.error(`[${category}]`, ...args);
  }
}

/**
 * Log debug timing information if the category is enabled
 * @param {string} category - Debug category name
 * @param {string} label - Timer label
 * @returns {function} - Function to call to end timing
 */
export function debugTime(category, label) {
  if (isDebugEnabled(category)) {
    const fullLabel = `[${category}] ${label}`;
    console.time(fullLabel);
    return () => console.timeEnd(fullLabel);
  }
  return () => {}; // No-op
}

/**
 * Get all enabled debug flags
 * @returns {string[]} - Array of enabled debug category names
 */
export function getEnabledDebugFlags() {
  return Object.keys(DEBUG_FLAGS).filter(key => DEBUG_FLAGS[key]);
}

/**
 * Print debug configuration summary
 */
export function printDebugConfig() {
  const enabled = getEnabledDebugFlags();
  if (enabled.length === 0) {
    console.log('[DEBUG] No debug flags enabled');
  } else if (DEBUG_ALL) {
    console.log('[DEBUG] ALL debug flags enabled');
  } else {
    console.log('[DEBUG] Enabled flags:', enabled.join(', '));
  }
}

// Export individual flags for conditional rendering
export const DEBUG = DEBUG_FLAGS;

export default {
  isDebugEnabled,
  debugLog,
  debugWarn,
  debugError,
  debugTime,
  getEnabledDebugFlags,
  printDebugConfig,
  DEBUG,
};
